File: plugin/src/main/java/io/github/kdy05/physicalFighters/ability/impl/Medic.java
package io.github.kdy05.physicalFighters.ability.impl;

import io.github.kdy05.physicalFighters.ability.Ability;
import io.github.kdy05.physicalFighters.ability.AbilitySpec;
import io.github.kdy05.physicalFighters.ability.AbilityUtils;
import io.github.kdy05.physicalFighters.game.EventManager;
import io.github.kdy05.physicalFighters.util.EventData;
import io.github.kdy05.physicalFighters.util.SoundUtils;
import org.bukkit.ChatColor;
import org.bukkit.entity.LivingEntity;
import org.bukkit.entity.Player;
import org.bukkit.event.Event;
import org.bukkit.event.entity.EntityDamageByEntityEvent;
import org.bukkit.event.player.PlayerInteractEvent;

import java.util.UUID;

public final class Medic extends Ability {
    public Medic(UUID playerUuid) {
        super(AbilitySpec.builder("메딕", Type.ActiveImmediately, Rank.B)
                .cooldown(8)
                .guide(Usage.IronAttack + "타인의 체력을 6만큼 회복합니다.",
                        Usage.IronRight + "자신의 체력을 6만큼 회복합니다.")
                .build(), playerUuid);
    }

    @Override
    public void registerEvents() {
        EventManager.registerEntityDamageByEntity(new EventData(this));
        registerRightClickEvent();
    }

    @Override
    public int checkCondition(Event event, int CustomData) {
        if (CustomData == 0) {
            EntityDamageByEntityEvent event0 = (EntityDamageByEntityEvent) event;
            if (event0.getEntity() instanceof LivingEntity &&
                    isOwner(event0.getDamager()) && isValidItem(Ability.DefaultItem)) {
                return 0;
            }
        } else if (CustomData == 1) {
            PlayerInteractEvent event1 = (PlayerInteractEvent) event;
            if (isOwner(event1.getPlayer()) && isValidItem(Ability.DefaultItem)) {
                return 1;
            }
        }
        return -1;
    }

    @Override
    public void applyEffect(Event event, int CustomData) {
        if (CustomData == 0) {
            EntityDamageByEntityEvent event0 = (EntityDamageByEntityEvent) event;
            LivingEntity entity = (LivingEntity) event0.getEntity();
            AbilityUtils.healEntity(entity, 6);
            if (getPlayer() == null) return;
            entity.sendMessage(String.format(ChatColor.GREEN
                    + "%s의 메딕 능력으로 체력을 6 회복했습니다.", getPlayer().getName()));
            SoundUtils.playSuccessSound(getPlayer());
            getPlayer().sendMessage(String.format(ChatColor.GREEN
                    + "%s의 체력을 6 회복시켰습니다.", entity.getName()));
            event0.setCancelled(true);
        } else if (CustomData == 1) {
            PlayerInteractEvent event1 = (PlayerInteractEvent) event;
            Player p2 = event1.getPlayer();
            AbilityUtils.healEntity(p2, 6);
            SoundUtils.playSuccessSound(getPlayer());
            p2.sendMessage(ChatColor.GREEN + "자신의 체력을 6 회복했습니다.");
        }
    }
}

File: abilities/src/main/kotlin/io/github/kdy05/physicalFighters/ability/impl/Medic.kt
package io.github.kdy05.physicalFighters.ability.impl

import io.github.kdy05.abilityAPI.ability.AbilityMeta
import io.github.kdy05.abilityAPI.rank.Rank
import io.github.kdy05.abilityAPI.skill.SkillContext
import io.github.kdy05.physicalFighters.ability.PFAbility
import io.github.kdy05.physicalFighters.ability.AbilityUtils
import io.github.kdy05.physicalFighters.util.SoundUtils
import org.bukkit.ChatColor
import org.bukkit.Material
import org.bukkit.entity.LivingEntity
import org.bukkit.entity.Player

@AbilityMeta(
    name = "메딕",
    rank = Rank.B,
    guide = [
        "(철괴 공격) 타인의 체력을 6만큼 회복합니다.",
        "(철괴 우클릭) 자신의 체력을 6만큼 회복합니다."
    ]
)
class Medic(owner: Player, context: SkillContext) : PFAbility(owner, context) {
    override fun buildSkills() = listOf(object : ActiveSkill() {
        override val cooldownTicks = 160L
        private var pendingAction: (() -> Unit)? = null

        override fun register() {
            onEntityDamage { e ->
                if (owner.inventory.itemInMainHand.type != Material.IRON_INGOT) return@onEntityDamage
                val entity = e.entity as? LivingEntity ?: return@onEntityDamage
                pendingAction = {
                    AbilityUtils.healEntity(entity, 6.0)
                    entity.sendMessage("${ChatColor.GREEN}${owner.name}의 메딕 능력으로 체력을 6 회복했습니다.")
                    SoundUtils.playSuccessSound(owner)
                    owner.sendMessage("${ChatColor.GREEN}${entity.name}의 체력을 6 회복시켰습니다.")
                    e.isCancelled = true
                }
                activate()
            }
            onRightClick {
                if (owner.inventory.itemInMainHand.type != Material.IRON_INGOT) return@onRightClick
                pendingAction = {
                    AbilityUtils.healEntity(owner, 6.0)
                    SoundUtils.playSuccessSound(owner)
                    owner.sendMessage("${ChatColor.GREEN}자신의 체력을 6 회복했습니다.")
                }
                activate()
            }
        }

        override fun onActivate() {
            pendingAction?.invoke()
            pendingAction = null
        }
    })
}
